berdi/src/main/java/com/di/berdi/Injector.kt

```diff
@@ -0,0 +1,69 @@
+package com.di.berdi
+
+import android.content.Context
+import com.di.berdi.annotation.Qualifier
+import kotlin.reflect.KFunction
+import kotlin.reflect.KParameter
+import kotlin.reflect.full.declaredFunctions
+import kotlin.reflect.full.hasAnnotation
+import kotlin.reflect.full.valueParameters
+import kotlin.reflect.jvm.jvmErasure
+
+class Injector(private val container: Container, private val context: Context) {
+
+    fun inject(module: Module) {
+        val functions = module::class.declaredFunctions
+        functions.forEach { func -> recursive(functions, func, module) }
+    }
+
+    private fun recursive(
+        functions: Collection<KFunction<*>>,
+        func: KFunction<*>,
+        module: Module,
+    ) {
+        val funcClazz = func.returnType.jvmErasure
+        val annotation = getQualifierAnnotation(func)
+
+        // 이미 생성했다면 return
+        if (container.getInstance(funcClazz, annotation) != null) return
+
+        // 파라미터가 없다면 만든다
+        if (func.valueParameters.isEmpty()) {
+            func.call(module)?.let { container.setInstance(it, funcClazz, annotation) }
+            return
+        }
+
+        // 파라미터를 하나씩 채운다
+        val params = func.valueParameters.map { param ->
+            when (param.type.jvmErasure) {
+                Context::class -> context
+                else -> getParamInstance(param, functions, module)
+            }
+        }
+
+        // 생성하고 컨테이너 에 삽입
+        val newInstance = requireNotNull(func.call(module, *params.toTypedArray()))
+        container.setInstance(newInstance, funcClazz, annotation)
+    }
+
+    private fun getQualifierAnnotation(func: KFunction<*>): Annotation? =
+        func.annotations.firstOrNull { it.annotationClass.hasAnnotation<Qualifier>() }
```

Comment:
```suggestion
    private val KFunction<*>.qualifierAnnotation: Annotation? = 
        this.annotations.firstOrNull { it.annotationClass.hasAnnotation<Qualifier>() }
     private val KAnnotatedElement.qualifierAnnotation: Annotation? = 
        this.annotations.firstOrNull { it.annotationClass.hasAnnotation<Qualifier>() }
```
레벨 1때에는 내부 확장함수는 지양해야하는 기술로 인식 되었는데, 확장함수로 가독성을 높일 수 있다고 생각하고 있습니다.
추가로 DI 라이브러리에서 지원하는 확장함수로 Util로 완전히 밖으로 빼도 괜찮다고 생각합니다.
베르는 이에 대해서 어떤 의견을 가지고 있는지 궁금합니다!

Reply:
정말 좋네요! 하지만 라이브러리 사용자는 사용할 수 없도록 제한하겠습니다!